### DISCUSSION TITLE
General spring between shaft and disc

### DISCUSSION THREAD

**Discussion by PetterTh** [petrobras/ross#1088] [Ideas]
Hi

I'm currently working on a project where I need to include the flexibility of the structure between my generatorrotor and the shaft. As far as I can see, this is not possible within ROSS today.

In this case both the direct stiffness $k_{xx}$,$k_{yy}$ and the direct bending $k_{\alpha \alpha}$ and $k_{\beta \beta}$ would be needed.
Is this a possible feature to include in ROSS?

**Comment by ViniciusTxc3** [comment]
Hi @PetterTh, could you detail more your case, so I can understand it better and try to help you create this application?

**Reply by PetterTh** [reply]
Hi

Sorry for the late reply.

In my particular case I have a generator where the mass is located far from the shaft. The generator mass is connected to the shaft with quite slender beams. These beams cannot be assumed to be infinitely stiff, so I would need to somehow model their flexibility in atleast 4 DOFs. 

So that is why I asked for this spring. To model these connecting beams.
I hope this clarified my request.

Petter

**Reply by jguarato** [reply]
Maybe the `CouplingElement` could work for what you need. You need to create it and put it in the `shaft_elements` list.

**Reply by PetterTh** [reply]
Thank you @jguarato 
I've given this some tests, and I'm not getting quite the results I'm hoping for.

I have 0 mass and momentum on the left node and everything on the right. 
I set n= the node where I want it to connect.
```
GeneratorModel=rs.CouplingElement(m_l=0,m_r=gMass,Ip_l=0,Ip_r=gIp,Id_l=0,Id_r=gId,
                                  kt_x=gen_rad_stiffness,kt_y=gen_rad_stiffness,kt_z=gen_ax_stiffness,
                                  kr_x=gen_rot_stiffness_xy,kr_y=gen_rot_stiffness_xy,kr_z=gen_rot_stiffness_z,
                                  n=GenNode,o_d=7.5,L=0.25,scale_factor=10)
```
                                  
However the results does not seem to vary with changes in the stiffness values.
Having them all at 0 or 1e10 gives the same result.

Can it be due to n_r being set to n+1?      
```
        self.n = n
        self.n_l = n
        self.n_r = None
        if n is not None:
            self.n_r = n + 1
```
I want the right node to be only the rotor masses and not connected to another node in the shaft.

Do you know If there is something I can do to fix this?

**Reply by jguarato** [reply]
Sorry for the delay. Perhaps it wasn’t very clear regarding the arguments of the `CouplingElement` class. The parameters `m_l`, `m_r`, `Ip_l`, and `Ip_r` are not the properties of what is on the left or right of the coupling, but rather the properties of the coupling element itself.

You can think of it as if the coupling is divided into two parts: the mass of the left part of the coupling is `m_l`, and the mass of the right part is `m_r`. When you create this element, you need to connect it to other shaft elements, as shown in the ROSS documentation: [Creating CouplingElement](https://ross.readthedocs.io/en/stable/user_guide/tutorial_part_1.html#creating-coupling-element-couplingelement-class)

**Example:**
```python
import ross as rs
from ross.material import steel

# Define shaft elements (only on the right side of coupling)
shaft_right = [
    rs.ShaftElement(
        L=0.25,
        idl=0.0,
        odl=0.05,
        material=steel,
        shear_effects=True,
        rotary_inertia=True,
        gyroscopic=True
    ),
    rs.ShaftElement(
        L=0.25,
        idl=0.0,
        odl=0.05,
        material=steel,
        shear_effects=True,
        rotary_inertia=True,
        gyroscopic=True
    )
]

# Define the coupling element
Ip = 4
m = 150
coupling = rs.CouplingElement(
    m_l=m / 2,  # Assuming equal distribution
    m_r=m / 2, 
    Ip_l=Ip / 2,
    Ip_r=Ip / 2, 
    kr_z=1e6,
)

# Combine elements into rotor model: coupling first (left side empty) then shaft_right
shaft_elements = [coupling] + shaft_right

rotor = rs.Rotor(
    shaft_elements=shaft_elements,
)
```

**Reply by PetterTh** [reply]
Hi @jguarato 

Thank you so much for taking the time to answer my questions.

I don't think my problem is solved by your example.
I may not have given the best explanation of my setup.
Below is a sketch of my shaft-line.

Shaft: Black
Turbine: Blue
Generator Masses: Grey
Generator Arms: Yellow
Bearings: Red

What I would like to do, is to include the stiffness of the generator arms. In some projects they are quite weak and the generator cannot be modeled as an infinitely stiff disc. There should be some flexibility between the generator masses and shaft.

Can this be solved with the Coupling element?
I've tried to model this as a co-axial multirotor system. But I could not get the coupling element to work in between the rotors.
(Btw: I tried to look at the co-axial example: https://ross.readthedocs.io/en/stable/user_guide/example_25.html, but I can't see anything about co-axiality there)

<img width="1147" height="392" alt="image" src="https://github.com/user-attachments/assets/0367393f-875f-460d-99cc-047db16e8bc5" />

**Reply by jguarato** [reply]
I agree with your reasoning — in this case, it is indeed more appropriate to use **shaft elements** instead of a **coupling element**.

In ROSS, you can **place multiple shaft elements at the same node**, which allows you to represent different geometric sections or materials that meet at a common interface.

To achieve the desired stiffness, you can:

- **Use multiple shaft elements and adjust the material properties** of the shaft elements (e.g., modify the Young’s modulus *E* or shear modulus *G* to match experimental stiffness), or
- **Use a single shaft element** at that node with **equivalent diameters and lengths** that represent the geometry of the physical section being modeled.
    - See page 158 of this reference: [*A Handbook on Torsional Vibration*](https://ia902909.us.archive.org/11/items/in.ernet.dli.2015.140286/2015.140286.A-Handbook-On-Torsional-Vibration.pdf)

Example illustrating how to use multiple elements:
```python
import ross as rs

steel = rs.materials.steel

od_shaft = 0.08
od_arm = 0.5
od_gen = 0.6

# main shaft
L = [0.1, 0.25, 0.1, 0.1, 0.15, 0.2, 0.8, 0.4]
shaft = [
    rs.ShaftElement(
        L=l,
        idl=0.0,
        odl=od_shaft,
        idr=0.0,
        odr=od_shaft,
        material=steel,
        shear_effects=True,
        rotary_inertia=True,
        gyroscopic=True,
    )
    for l in L
]

# generator arm (yellow)
n = 3
mat_arm = rs.Material("Mat_arm", E=211e9, G_s=81.2e9, rho=7810, color="yellow")
gen_arm = rs.ShaftElement(
    n=n,
    L=L[n],
    idl=od_shaft,
    odl=od_arm,
    idr=od_shaft,
    odr=od_arm,
    material=mat_arm,
    shear_effects=True,
    rotary_inertia=True,
    gyroscopic=True,
)

# generator masses (gray)
gen_mass = []
mat_gen = rs.Material("Mat_mass", E=211e9, G_s=81.2e9, rho=7810, color="brown")
for n in [2, 3, 4]:
    el = rs.ShaftElement(
        n=n,
        L=L[n],
        idl=od_arm,
        odl=od_gen,
        idr=od_arm,
        odr=od_gen,
        material=mat_gen,
        shear_effects=True,
        rotary_inertia=True,
        gyroscopic=True,
    )
    gen_mass.append(el)

shaft_elements = [*shaft, gen_arm, *gen_mass]

turbine = rs.DiskElement(n=8, m=600, Id=3, Ip=6)

kxx = 1e6
kyy = 0.8e6
kzz = 0.1e6
brg1 = rs.BearingElement(n=1, kxx=kxx, kyy=kyy, kzz=kzz, cxx=0, cyy=0, czz=0)
brg2 = rs.BearingElement(n=6, kxx=kxx, kyy=kyy, kzz=kzz, cxx=0, cyy=0, czz=0)
brg3 = rs.BearingElement(n=7, kxx=kxx, kyy=kyy, kzz=kzz, cxx=0, cyy=0, czz=0)

bearing_elements = [brg1, brg2, brg3]

rotor = rs.Rotor(shaft_elements=shaft_elements, disk_elements=[turbine], bearing_elements=bearing_elements, tag="generator_sketch")

rotor.plot_rotor().show()
```
<img width="614" height="525" alt="newplot" src="https://github.com/user-attachments/assets/49716895-bf7a-484c-bb35-f21fc70a8f48" />

**Comment by PetterTh** [comment]
Excellent! Thank you!

I'll give this a try.

One question:

I'm not able to see how you in the setup tell which parts are connected and which are not. 

It seems to me we have the following connections:
Node:
2: Shaft and generator mass
3: Shaft, arms and generator mass
4: Shaft, arms and generator mass
5: Shaft and generator mass

When both the shaft and the generator mass both connect to node 2 (and 5), will they also share degrees of freedom here? Like they are rigidly connected at this node.
Or Does ross take into accound the diameters to see which shaft parts are connected?

**Reply by jguarato** [reply]
Elements built using the same node share the same degrees of freedom.
Thus, the displacement and rotation at that node are common to all connected elements.

When assembling the global rotor matrices, ROSS automatically combines the stiffness, mass, damping, and gyroscopic contributions of all elements connected at the same node.

The diameters do not affect the connectivity between elements — they only influence each element’s mass and stiffness.

**Reply by PetterTh** [reply]
Ok, then I assume that the example you gave above will not work as intended?
Since the generator mass will be directly coupled to the shaft at node 2 through 5.

**Comment by joselivera** [comment]
I actually have a similar problem, described in the topic:

https://github.com/petrobras/ross/discussions/1227

Thanks for the shaft in the same node trick @jguarato.
I'll take a look in the approach taken here.
